api/src/workers/base-event-handler.ts

```diff
@@ -1,129 +1,95 @@
-import { BlockCursorService } from "../services/block-cursor";
-import { FlowService } from "../services/flow";
 import * as fcl from "@onflow/fcl";
 import { send } from "@onflow/sdk-send";
 import { getEvents } from "@onflow/sdk-build-get-events";
-import { BlockCursor } from "../models/block-cursor";
 
-interface EventDetails {
-  blockHeight: number;
-}
+import { BlockCursor } from "../models/block-cursor";
+import { BlockCursorService } from "../services/block-cursor";
+import { FlowService } from "../services/flow";
 
 // BaseEventHandler will iterate through a range of block_heights and then run a callback to process any events we
 // are interested in. It also keeps a cursor in the database so we can resume from where we left off at any time.
 abstract class BaseEventHandler {
-  private stepSize: number = 1000;
   private stepTimeMs: number = 1000;
-  private blockThreshold: number = 100;
+
   protected constructor(
     private readonly blockCursorService: BlockCursorService,
     private readonly flowService: FlowService,
-    private readonly eventName: string
+    public eventNames: string[]
   ) {}
 
   async run() {
     console.log("fetching latest block height");
+
     let latestBlockHeight = await this.flowService.getLatestBlockHeight();
+
     console.log("latestBlockHeight =", latestBlockHeight.height);
-    // create a cursor on the database
-    let blockCursor = await this.blockCursorService.findOrCreateLatestBlockCursor(
-      this.eventName,
-      latestBlockHeight.height
-    );
 
-    if (!blockCursor) {
-      throw new Error(
-        `invalid state, no block cursor defined for event_name=${this.eventName}`
+    const cursors = this.eventNames.map((eventName) => {
+      const cursor = this.blockCursorService.findOrCreateLatestBlockCursor(
+        latestBlockHeight.height,
+        eventName
       );
-    }
+      return { cursor, eventName };
+    });
 
-    // loop
-    let fromBlock, toBlock;
-
-    while (true) {
-      await this.sleep(this.stepTimeMs);
-      try {
-        // calculate fromBlock, toBlock
-        ({ fromBlock, toBlock } = await this.getBlockRange(blockCursor));
-      } catch (e) {
-        console.warn("error retrieving block range");
-        continue;
-      }
-
-      // make sure we query the access node only when we have a substantial gap
-      // between from and to block ranges; currently if we query the access node
-      // when the gap is too narrow, it returns an error.
```

Comment:
I think this might be happening again. I'm getting this error fairly frequently:


> Error retrieving events for block range fromBlock=26996926 toBlock=26996928 Error: 
start height 26996926 is greater than the last sealed block height 26996925

I think it probably makes sense to bring back `blockThreshold`. However, you can reduce it to something smaller than 100. I think 5 or 10 would be fine.

Reply:
It will query from the block where it last saw an event it cares about, up to the latest block, so we don't need a threshold. We could set a limit ... say, if the block range is more than 1000, than move the window to 1000 from the latest block.